pkg/scalers/neo4j_scaler.go

```diff
@@ -0,0 +1,210 @@
+	scalerIndex          int
+}
+
+func (s *neo4jScaler) GetMetricSpecForScaling(context.Context) []v2.MetricSpec {
+	externalMetric := &v2.ExternalMetricSource{
+		Metric: v2.MetricIdentifier{
+			Name: GenerateMetricNameWithIndex(s.metadata.scalerIndex, s.metadata.metricName),
+		},
+		Target: GetMetricTarget(s.metricType, s.metadata.queryValue),
+	}
+	metricSpec := v2.MetricSpec{
+		External: externalMetric, Type: externalMetricType,
+	}
+	return []v2.MetricSpec{metricSpec}
+}
+
+func (s *neo4jScaler) Close(ctx context.Context) error {
+	if s.driver != nil {
+		err := s.driver.Close(ctx)
+		if err != nil {
+			s.logger.Error(err, fmt.Sprintf("failed to close neo4j connection, because of %v", err))
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *neo4jScaler) getQueryResult(ctx context.Context) (int64, error) {
+	session := s.driver.NewSession(ctx, neo4j.SessionConfig{
+		AccessMode: neo4j.AccessModeWrite,
+	})
+	defer session.Close(ctx)
+	result, err := session.ExecuteWrite(ctx, matchItemFn(s.metadata.query, ctx))
+	if err != nil {
+		s.logger.Error(err, fmt.Sprintf("Couldn't execute query string because of %v", err))
+		return 0, err
+	}
+	res, err := strconv.ParseInt((fmt.Sprintf("%v", result)), 10, 64)
+	if err != nil {
+		s.logger.Error(err, fmt.Sprintf("Couldn't parse to int because of %v", err))
+		return 0, err
+	}
+	return res, nil
+}
+
+func matchItemFn(query string, ctx context.Context) neo4j.ManagedTransactionWork {
+	return func(tx neo4j.ManagedTransaction) (any, error) {
+		records, err := tx.Run(ctx, query, nil)
+		if err != nil {
+			return nil, err
+		}
+		if records.Peek(ctx) {
+			record := records.Record()
+			if err != nil {
+				return nil, err
+			}
+			return record.Values[1], nil
+		}
+
+		return 0, nil
+	}
+}
+
+func parseNeo4jMetadata(config *ScalerConfig) (*neo4jMetadata, string, error) {
+	var connStr string
+	meta := neo4jMetadata{}
+
+	if val, ok := config.TriggerMetadata["query"]; ok {
+		meta.query = val
+	} else {
+		return nil, "", fmt.Errorf("no query given")
+	}
+
+	if val, ok := config.TriggerMetadata["queryValue"]; ok {
+		queryValue, err := strconv.ParseInt(val, 10, 64)
+		if err != nil {
+			return nil, "", fmt.Errorf("failed to convert %v to int, because of %v", val, err.Error())
+		}
+		meta.queryValue = queryValue
+	} else {
+		return nil, "", fmt.Errorf("no queryValue given")
+	}
+
+	meta.activationQueryValue = 0
+	if val, ok := config.TriggerMetadata["activationQueryValue"]; ok {
+		activationQueryValue, err := strconv.ParseInt(val, 10, 64)
+		if err != nil {
+			return nil, "", fmt.Errorf("failed to convert %v to int, because of %v", val, err.Error())
+		}
+		meta.activationQueryValue = activationQueryValue
+	}
+
+	switch {
+	case config.AuthParams["connectionString"] != "":
+		meta.connectionString = config.AuthParams["connectionString"]
+	case config.TriggerMetadata["connectionStringFromEnv"] != "":
+		meta.connectionString = config.ResolvedEnv[config.TriggerMetadata["connectionStringFromEnv"]]
```

Comment:
I haven't seen in the docs neo4j taking username and password from the connection string

Reply:
in that case, do we need the connection string? current implementation doesn't parse the username and password in case of providing connection string indeed, if we maintain connection string for providing `protocol://address:port`, we have to ensure that the username and password is already parsed.
In the same way, I guess we should accept and parse also the protocol (or at least supporting it with default value) because AFAI can see, there are other protocols already supported by the sdk

Reply:
Yes, they support one other protocol too. I can add a parameter for protocol in metadata too. Do you want me to remove connection string from the metadata?

Reply:
I guess that in this case, we can do both, remove the connection string and support protocol. `connectionString` sounds like it includes all the login information but in this case, you have to provide username and password even thought you are providing the connection string too 🤔

Reply:
Sure, I will try by removing `connectionString` and add protocol

Reply:
It worked